File: p4-heap.py
def heap_adjust(nums, start, end):
    if start>=end:
        return
    k=start
    sd=nums[k]
    i=2*k+1
    while i<=end:
        if i<end and nums[i]<nums[i+1]:
            i+=1
        if nums[i]>sd:
            nums[k] = nums[i]
            k=i
        else:
            break
        i=i*2+1
    nums[k]=sd        

def head_build(nums,length):
    for i in range((length-2)//2,-1,-1):
        heap_adjust(nums,i,length-1)
        
    
class Solution:
    def findKthLargest(self, nums, k: int) -> int:
        length = len(nums)
        head_build(nums,length)
        for i in range(0,k):
            tail = length-1-i
            t = nums[0]
            nums[0] = nums[tail]
            nums[tail] = t
            heap_adjust(nums,0,tail-1)

        return nums[length-1-(k-1)]
    
a= [9,4,6,0,8]
s=Solution()
s.findKthLargest(a,1)

File: cpp/p4-heap.cpp
#include <iostream>
#include <vector>
using namespace std;
// 这题有点脑残，必须升序排序，面向测试例AC

class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        buidlHeap(nums, 0, nums.size() - 1);
        int tmp;
        for (int i = 0; i < k; i++)
        {
            tmp = nums[0];
            nums[0] = nums[nums.size() - 1 - i];
            nums[nums.size() - 1 - i] = tmp;
            adjustHeap(nums, 0, nums.size() - 2 - i);
        }
        return nums[nums.size() - k];
    }

    int adjustHeap(vector<int> &nums, int lf, int rt)
    {
        int tmp = nums[lf];
        int k = lf;
        int i = 2 * k + 1; // 1打头时是2k ，0打头时是2k+1
        while (i <= rt)
        {
            if (i + 1 <= rt && nums[i + 1] > nums[i])
            {
                i++;
            }
            if (nums[i] > tmp)
            {
                nums[k] = nums[i];
                k = i;
                i = i * 2 + 1;
            }
            else
            {
                break;
            }
        }
        nums[k] = tmp;
        return 0;
    }

    int buidlHeap(vector<int> &nums, int lf, int rt)
    {
        for (int i = (rt - 1) / 2; i >= lf; i--)
        {
            adjustHeap(nums, i, rt);
        }
        return 0;
    }
};
